### PR TITLE
Plumbing for reproducing using new fuzzing engine pipeline.

### PR DESCRIPTION
**Pull request by oliverchang**
Part of #483.

- Move testcase_manager to bot.tasks.
- Introduce testcase_manager.TestcaseRunner to share more code between
  test_for_reproducibility and test_for_crash_with_retries.
- Add target name as first argument of engine crashes.

### CODE REVIEW

**Comment by oliverchang** [comment]
PTAL. I still need to write more tests, and some parts may be a little rough.

File: src/local/butler/lint.py [hunk 1]
```diff
@@ -203,7 +203,8 @@ def execute(_):
     futurize_command = 'futurize -0 {excludes} {file_path}'.format(
         excludes=futurize_excludes, file_path=file_path)
     futurize_output = _execute_command_and_track_error(futurize_command)
-    if 'No changes to ' not in futurize_output:
+    if ('No changes to ' not in futurize_output and
+        'No files need to be modified' not in futurize_output):
```
Comment (Dor1s) [inline src/local/butler/lint.py]: nit: is this change needed for this CL?

File: src/python/bot/tasks/testcase_manager.py [hunk 2]
```diff
@@ -487,37 +490,56 @@ def run_testcase_and_return_result_in_queue(crash_queue,
...
+class TestcaseRunner(object):
+  """Testcase runner."""
+
+  def __init__(self,
+               fuzzer_name,
+               testcase_path,
+               test_timeout,
+               gestures,
+               needs_http=False):
+    self._testcase_path = testcase_path
```
Comment (Dor1s) [inline src/python/bot/tasks/testcase_manager.py]: super nit: the order here matches neither the order of arguments above nor the alphabetical order

File: src/python/bot/tasks/testcase_manager.py [hunk 3]
```diff
@@ -526,121 +548,144 @@ def test_for_crash_with_retries(testcase,
...
               round_number=round_number),
           output=output)
-      continue
 
+    return crash_result
+
+  def _pre_run_cleanup(self):
+    """Common cleanup before running a testcase."""
+    # Cleanup any existing application instances and user profile directories.
+    # Cleaning up temp clears user profile directories and should be done before
```
Comment (Dor1s) [inline src/python/bot/tasks/testcase_manager.py]: nit: "clears" seems redundant

File: src/python/bot/tasks/testcase_manager.py [hunk 4]
```diff
@@ -526,121 +548,144 @@ def test_for_crash_with_retries(testcase,
...
+    """Try reproducing a crash with retries."""
+    self._pre_run_cleanup()
+    crash_result = None
 
-    if crash_result.is_security_issue() != testcase.security_flag:
-      logs.log('Crash security flag does not match, ignored.')
-      continue
+    for round_number in range(1, retries + 1):
+      crash_result = self.run(round_number)
+      state = self._get_crash_state(round_number, crash_result)
```
Comment (Dor1s) [inline src/python/bot/tasks/testcase_manager.py]: can move this to line 604

File: src/python/bot/tasks/variant_task.py [hunk 5]
```diff
@@ -36,7 +36,7 @@ def execute_task(testcase_id, job_type):
     # otherwise we will run into exceptions.
     return
 
-  # Setup testcase and its dependencies.
+  # Setup testcase and its duependencies.
```
Comment (Dor1s) [inline src/python/bot/tasks/variant_task.py]: nit: unintentional typo in `duependencies`

File: src/python/platforms/android/device.py [hunk 6]
```diff
@@ -452,7 +452,7 @@ def update_build(apk_path, force_update=True):
   # advance so that we do not have to write this to the device multiple
   # times.
   # TODO(mbarbella): Platforms code should not depend on fuzzing.
-  from fuzzing import testcase_manager
+  from bot.tasks import testcase_manager
```
Comment (Dor1s) [inline src/python/platforms/android/device.py]: Remove or fix the comment on the line 454?

File: src/python/tests/core/fuzzing/testcase_manager_test.py [hunk 7]
```diff
@@ -334,3 +336,126 @@ def test_end_marker(self):
...
+    ])
+
+  def test_test_for_reproducibility_blackbox(self):
+    """Test test_for_reproducibility (blackbox)."""
+    self.mock.get.return_value = None
+    self.mock.run_process.return_value = (1, 1, 'crash')
+    self.stack_analyzer_state.crash_state = 'state'
+    self.stack_analyzer_state.crash_type = 'Null-dereference'
+    result = testcase_manager.test_for_reproducibility(
+        'engine',
```
Comment (Dor1s) [inline src/python/tests/core/fuzzing/testcase_manager_test.py]: nit: change "engine" to some blackbox thing, like "dom_fuzzer", just to avoid any potential confusion

File: src/python/tests/core/fuzzing/testcase_manager_test.py [hunk 8]
```diff
@@ -334,3 +336,126 @@ def test_end_marker(self):
...
+        '/fuzz-testcase',
+        'state',
+        expected_security_flag=False,
+        test_timeout=10,
+        http_flag=False,
```
Comment (Dor1s) [inline src/python/tests/core/fuzzing/testcase_manager_test.py]: Would it make sense to try with `http_flag=True`, or it's out of scope here?

File: src/python/tests/core/fuzzing/testcase_manager_test.py (hunk 7 above)
Comment (inferno-chromium) [reply src/python/tests/core/fuzzing/testcase_manager_test.py]: +1

File: src/python/tests/core/fuzzing/testcase_manager_test.py [hunk 9]
```diff
@@ -334,3 +336,126 @@ def test_end_marker(self):
...
+        gestures=None)
+    self.assertTrue(result)
+    self.assertEqual(2, self.mock.run_process.call_count)
```
Comment (inferno-chromium) [inline src/python/tests/core/fuzzing/testcase_manager_test.py]: What does 2 signify, maybe added a comment.

File: src/python/bot/tasks/testcase_manager.py [hunk 10]
```diff
@@ -526,121 +548,144 @@ def test_for_crash_with_retries(testcase,
...
-                                                               round_number))
-  return False
+def test_for_reproducibility(fuzzer_name, testcase_path, expected_state,
+                             expected_security_flag, test_timeout, http_flag,
+                             gestures):
+  """Test to see if a crash is fully reproducible or is a one-time crasher."""
+  runner = TestcaseRunner(fuzzer_name, testcase_path, test_timeout, gestures,
+                          http_flag)
+  crash_retries = environment.get_value('CRASH_RETRIES')
+  return runner.test_reproduce_reliability(crash_retries, expected_state,
```
Comment (inferno-chromium) [inline src/python/bot/tasks/testcase_manager.py]: s/test_reproduce_reliability/reproduce_for_reliability 
to match style of reproduce_with_retries

File: src/python/bot/tasks/testcase_manager.py [hunk 11]
```diff
@@ -487,37 +490,56 @@ def run_testcase_and_return_result_in_queue(crash_queue,
...
+      # Read target_name + args from flags file.
+      additional_command_line_flags = get_additional_command_line_flags(
+          testcase_path)
+      self._arguments = additional_command_line_flags.split()
+      target_name = self._arguments.pop(0)
+
+      build_dir = environment.get_value('BUILD_DIR')
+      self._target_path = engine_common.find_fuzzer_path(build_dir, target_name)
+      if not self._target_path:
+        raise RuntimeError('Failed to find target ' + target_name)
```
Comment (inferno-chromium) [inline src/python/bot/tasks/testcase_manager.py]: This is bad for regression testing, can happen in cases where target is not there in old builds, see comment at end of find_fuzzer_path. We need to find a way to bail out and not retry.

File: src/python/bot/tasks/testcase_manager.py [hunk 12]
```diff
@@ -526,121 +548,144 @@ def test_for_crash_with_retries(testcase,
...
+def test_for_crash_with_retries(testcase,
+                                testcase_path,
+                                test_timeout,
+                                http_flag=False,
+                                compare_crash=True):
+  """Test for a crash and return crash parameters like crash type, crash state,
+  crash stacktrace, etc."""
+  runner = TestcaseRunner(testcase.fuzzer_name, testcase_path, test_timeout,
+                          testcase.gestures, http_flag)
+  crash_retries = environment.get_value('CRASH_RETRIES')
```
Comment (inferno-chromium) [inline src/python/bot/tasks/testcase_manager.py]: can we just initialize crash_retries in TestcaseRunner itself, i don't see we are setting anything custom in test_for_crash_with_retries or test_for_reproducibility

File: src/local/butler/lint.py (hunk 1 above)
Comment (oliverchang) [reply src/local/butler/lint.py]: Yes, as otherwise lint doesnt work.

File: src/local/butler/reproduce.py [hunk 13]
```diff
@@ -31,9 +31,9 @@
 from base import utils
 from bot.tasks import commands
 from bot.tasks import setup
+from bot.tasks import testcase_manager
```
[the remark this answers is not in the document]
Comment (oliverchang) [reply src/local/butler/reproduce.py]: I couldnt find another existing one, and tasks seemed suitable since its something thats only done in many of the tasks. There's also bot.fuzzers but that doesn't seem right either because it's not related to any particular fuzzer. Preferences here? I could also create a new subdir here like "fuzzing".

File: src/python/bot/tasks/testcase_manager.py (hunk 12 above)
Comment (oliverchang) [reply src/python/bot/tasks/testcase_manager.py]: I did this to make TestcaseRunner more general and not depend on more global state (i.e. an environment variable).

**Comment by inferno-chromium** [comment]
How about bot.testcase_management similar to build_management or even
bot.fuzzing is fine but don't like conflict with fuzzing package. I am fine
with original location if it makes it easy.

File: src/python/bot/tasks/testcase_manager.py (hunk 2 above)
Comment (oliverchang) [reply src/python/bot/tasks/testcase_manager.py]: Fixed to match order of arguments.

File: src/python/bot/tasks/testcase_manager.py (hunk 3 above)
Comment (oliverchang) [reply src/python/bot/tasks/testcase_manager.py]: Fixed.

File: src/python/bot/tasks/testcase_manager.py (hunk 4 above)
Comment (oliverchang) [reply src/python/bot/tasks/testcase_manager.py]: This logs the state even if we early exit in the following conditions, so we need this here.

File: src/python/bot/tasks/testcase_manager.py (hunk 10 above)
Comment (oliverchang) [reply src/python/bot/tasks/testcase_manager.py]: This is unchanged from before, and I'm not sure that reproduce_for_reliability is as clear. I think that "test_for_reproducibility" more clearly states the intention that we're seeing how reliable the reproducibility is?

File: src/python/bot/tasks/variant_task.py (hunk 5 above)
Comment (oliverchang) [reply src/python/bot/tasks/variant_task.py]: Fixed.

File: src/python/platforms/android/device.py (hunk 6 above)
Comment (oliverchang) [reply src/python/platforms/android/device.py]: Fixed the comment.

**Comment by oliverchang** [comment]
bot.testcase_managerment.testcase_manager sounds very redundant...

How about just bot.testcase_manager (i.e. top level import) ?

Btw, there's also other stuff in tasks that aren't exactly a task -- e.g. setup, trials etc.

File: src/local/butler/reproduce.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (oliverchang) [reply src/local/butler/reproduce.py]: Moved to bot.testcase_manager as discussed.

File: src/python/bot/tasks/testcase_manager.py (hunk 11 above)
Comment (oliverchang) [reply src/python/bot/tasks/testcase_manager.py]: Handled this exception in the callers.

File: src/python/bot/tasks/testcase_manager.py (hunk 10 above)
Comment (oliverchang) [reply src/python/bot/tasks/testcase_manager.py]: Leaving this comment open. Happy to rename everything if you still think reproduce_for_reliability is better!

File: src/python/tests/core/fuzzing/testcase_manager_test.py (hunk 7 above)
Comment (oliverchang) [reply src/python/tests/core/fuzzing/testcase_manager_test.py]: Done.

File: src/python/tests/core/fuzzing/testcase_manager_test.py (hunk 8 above)
Comment (oliverchang) [reply src/python/tests/core/fuzzing/testcase_manager_test.py]: I think that would be out of scope for this test. http_flag testing should go into a test for get_command_line_for_application.

File: src/python/tests/core/fuzzing/testcase_manager_test.py (hunk 9 above)
Comment (oliverchang) [reply src/python/tests/core/fuzzing/testcase_manager_test.py]: Added.

File: src/python/bot/tasks/testcase_manager.py (hunk 10 above)
Comment (inferno-chromium) [reply src/python/bot/tasks/testcase_manager.py]: ok lets keep-as is.

File: src/python/bot/tasks/testcase_manager.py (hunk 11 above)
Comment (inferno-chromium) [reply src/python/bot/tasks/testcase_manager.py]: Thanks!

File: src/python/bot/tasks/testcase_manager.py (hunk 12 above)
Comment (inferno-chromium) [reply src/python/bot/tasks/testcase_manager.py]: Sounds good, thanks!
